### DISCUSSION TITLE
how can I get raw data from request in custom endpoint

### DISCUSSION THREAD

**Discussion by dirums** [payloadcms/payload#3346] [General]
Hi team,
I wanted to know how I can access the raw incoming data (i.e. no JSON) in custome endpoint request.

**Comment by matthijs166** [comment]
Do you have an example and/or a use case?

**Reply by hugoburguete** [reply]
I know this is an old request but one example would be creating signatures from Stripe (https://docs.stripe.com/webhooks#example-endpoint). Having the rawBody is required for their official signing library

**Reply by andershermansen** [reply]
@hugoburguete Here is how you can do it with payload 3 and the stripe SDK using a payload endpoint handler.

```ts
import type { PayloadHandler } from 'payload';
import Stripe from 'stripe';

const stripeWebhookEndpointHandler: PayloadHandler = async (req) => {
    if (!req.arrayBuffer) {
        console.error('Webhook Error: No data');
        return Response.json({ error: 'Webhook Error' }, { status: 400 });
    }

    const arrayBuffer = await req.arrayBuffer();
    const rawBody = Buffer.from(arrayBuffer);

    const stripe = new Stripe(process.env.STRIPE_SECRET_KEY!);
    const webhookSecret: string = process.env.STRIPE_WEBHOOK_SECRET!;
    const signature = req.headers.get('stripe-signature')!;

    let event: Stripe.Event;
    try {
        event = stripe.webhooks.constructEvent(rawBody, signature, webhookSecret);
    } catch (err) {
        console.error('Webhook Error:', err);
        return Response.json({ error: 'Webhook Error' }, { status: 400 });
    }

    const { payload } = req;

    console.log('Stripe webhook event', event);

    // handle the event

    return Response.json({});
};

export default stripeWebhookEndpointHandler;
```

If you have a collection payments for example you can hook it up in the config like this:
```ts
    endpoints: [
        {
            path: '/webhook',
            method: 'post',
            handler: stripeWebhookEndpointHandler,
        },
    ],
```

and then you can register /api/payments/webhook in stripe as a webhook handler.

**Reply by andershermansen** [reply]
@hugoburguete I see now that you used stripe as an example on how to do it, not asked how to do it with stripe. The answer is good anyway for anybody else looking.

@dirums Here you see how to do it in an endpoint handler anyway, to get the raw body data.
